designs/mfa_backup_code.rst

```diff
@@ -0,0 +1,81 @@
+
+Account recovery (backup) mfa codes draft design
+------------------------
+MFA requires users to have access to a working physical device, however that might not be always possible.
+
+Backup mfa code allows:
+
+- single-device accounts to be restored
+- initial setup for passwordless-login
+
+Overview
+------------------------
+
+If the user has a Backup code, then they can replace the TOTP/WebAuthn challenge with it during login. For example, TOTP + password auth will be Backup code + password instead.
+
+To implement this, we can add Backup code as a new AuthCredential, so that it can be used in the auth process.
+
+To validate Backup code, we simply check if the input matches the Backup code saved in database, which should be hashed with salt (similar to how password is validated).
+
+A user can only have one Backup code, if user wants a new Backup code:
```

Comment:
I'd check how github does this - since the backup code is "once use" and get's scratched off (deleted) after use, then we also need:

* An async message to remove that backup code from the account after usage
* Multiple backup codes.

Github for example issues batches of 5 or 10 at a time that can be printed. I think we should do similar.

Reply:
Github issues a batch of 16 codes. When a code is used, it is no longer valid (i.e. you can only use the remaining 15 codes).
When user generates a new batch of backup codes, the previous codes will no longer be valid.

Reply:
One question though: Why do we need a separate async message? Couldn't we just do it in the same transaction?

Reply:
Yep, that's the thought I had is to have a batch of codes and you use one at a time. 

Well, the thing about the auth process is that currently while it's in a single transaction lock, that transaction is NOT a writeable transaction, and the intent is to make it so that authentications can proceed in parallel.

Because the server can also operate as a single thread, and there are a limited set of db tickets, that also prevents opening nested transactions too.

So the correct way here to perform the "write" to invalidate the code is to send a message to the async task queue - you can see how this is done for "on login pw upgrades". A similar process could be used here, but when the recovery code is used and is found correct, we immediately send the async task to deactivate it. Theoretically, this means there is a small window between "use of the backup code" and the async action being processed to actually cause the invalidation, but this window is short. So I think it's an acceptable compromise here.

Reply:
Yep I was wondering what if another agent login with the same code while the async action is being processed. 
Thanks for clearing that up!